schema/query/parse_test.go

```diff
@@ -0,0 +1,348 @@
+		},
+		{
+			`{"foo": {"$ne": {"bar": baz}}}`,
+			Query{},
+			errors.New("char 24: foo: $ne: unexpected char 'b'"),
+		},
+		{
+			`{"foo": {"$ne": {"bar": "baz"]}}`,
+			Query{},
+			errors.New("char 29: foo: $ne: expected '}' got ']'"),
+		},
+		{
+			`{"bar": {"$gt": "1"}}`,
+			Query{},
+			errors.New("char 16: bar: $gt: not a number"),
+		},
+		{
+			`{"bar": {"$gt": 1ee0}}`,
+			Query{},
+			errors.New("char 16: bar: $gt: not a number: strconv.ParseFloat: parsing \"1ee0\": invalid syntax"),
+		},
+
+		{
+			`{"bar": {"$in": {"bar": "1"}}}`,
+			Query{},
+			errors.New("char 16: bar: $in: expected '[' got '{'"),
+		},
+		{
+			`{"bar": {"$in": ["bar": "1"]}}`,
+			Query{},
+			errors.New("char 22: bar: $in: expected ',' or ']' got ':'"),
+		},
+		{
+			`{"bar": {"$in": [bar]}}`,
+			Query{},
+			errors.New("char 17: bar: $in: item #0: unexpected char 'b'"),
+		},
+		{
+			`{"bar": {"$in": "bar"}}`,
+			Query{},
+			errors.New("char 16: bar: $in: expected '[' got '\"'"),
+		},
+		{
+			`{"$or": "foo"}`,
+			Query{},
+			errors.New("char 8: $or: expected '[' got '\"'"),
+		},
+		{
+			`{"$or": ["foo", "bar"]}`,
+			Query{},
+			errors.New("char 9: $or: expected '{' got '\"'"),
+		},
+		{
+			`{"$or": [{"foo": "bar"}}`,
+			Query{},
+			errors.New("char 23: $or: expected ']' got '}'"),
+		},
+		{
+			`{"$or": []}`,
+			Query{},
+			errors.New("char 10: $or: two expressions or more required"),
+		},
+		{
+			`{"$or": [{"foo": "bar"}]}`,
+			Query{},
+			errors.New("char 24: $or: two expressions or more required"),
+		},
+		{
+			`{"$and": "foo"}`,
+			Query{},
+			errors.New("char 9: $and: expected '[' got '\"'"),
+		},
+		{
+			`{"$and": [{"foo": "bar"}]}`,
+			Query{},
+			errors.New("char 25: $and: two expressions or more required"),
+		},
+		{
+			`{"$and": ["foo", "bar"]}`,
+			Query{},
+			errors.New("char 10: $and: expected '{' got '\"'"),
+		},
+		{
+			`{"foo": {"$regex": "b[..?r"}}`,
+			Query{},
+			errors.New("char 27: foo: $regex: invalid regex: error parsing regexp: missing closing ]: `[..?r`"),
+		},
+		{
+			`{"foo": {"$regex": "b[a-z)r"}}`,
+			Query{},
+			errors.New("char 28: foo: $regex: invalid regex: error parsing regexp: missing closing ]: `[a-z)r`"),
+		},
+		{
+			`{"foo": {"$regex": "b(?=a)r"}}`,
+			Query{},
+			errors.New("char 28: foo: $regex: invalid regex: error parsing regexp: invalid or unsupported Perl syntax: `(?=`"),
+		},
+		// Hierarchy issues
+		{
+			`{"$ne": "bar"}`,
+			Query{},
+			errors.New("char 1: $ne: invalid placement"),
+		},
+		{
+			`{"$exists": true}`,
+			Query{},
+			errors.New("char 1: $exists: invalid placement"),
+		},
+		{
+			`{"$gt": 1}`,
+			Query{},
+			errors.New("char 1: $gt: invalid placement"),
+		},
+		{
+			`{"$in": [1,2]}`,
+			Query{},
+			errors.New("char 1: $in: invalid placement"),
+		},
+		{
+			`{"$regex": "someregexpression"}`,
+			Query{},
+			errors.New("char 1: $regex: invalid placement"),
+		},
+	}
+	for _, tt := range tests {
```

Comment:
Should this use sub-tests for better listing of the results, and maybe do them in parallel?


```go
 +	for i := range tests {
 +		tt := tests[i]   // capture sub-test.
 +		t.Run(strings.Replace(tt.query, " ", "", -1), func(t *testing.T) {
 +			t.Parallel()
 +			got, err := Parse(tt.query)
 +			if !reflect.DeepEqual(err, tt.err) {
...
 + 		})
 +	}
```